source/extensions/common/wasm/v8/v8.cc

```diff
@@ -0,0 +1,621 @@
+  void getFunction(absl::string_view function_name, T* f) override {                               \
+    getModuleFunctionImpl(function_name, f);                                                       \
+  };
+  FOR_ALL_WASM_VM_EXPORTS(_GET_MODULE_FUNCTION)
+#undef _GET_MODULE_FUNCTION
+
+private:
+  template <typename T>
+  std::unique_ptr<Global<T>> registerHostGlobalImpl(absl::string_view module_name,
+                                                    absl::string_view name, T initial_value);
+
+  template <typename... Args>
+  void registerHostFunctionImpl(absl::string_view module_name, absl::string_view function_name,
+                                void (*function)(void*, Args...));
+
+  template <typename R, typename... Args>
+  void registerHostFunctionImpl(absl::string_view module_name, absl::string_view function_name,
+                                R (*function)(void*, Args...));
+
+  template <typename... Args>
+  void getModuleFunctionImpl(absl::string_view function_name,
+                             std::function<void(Context*, Args...)>* function);
+
+  template <typename R, typename... Args>
+  void getModuleFunctionImpl(absl::string_view function_name,
+                             std::function<R(Context*, Args...)>* function);
+
+  wasm::vec<byte_t> source_ = wasm::vec<byte_t>::invalid();
+  wasm::own<wasm::Store> store_;
+  wasm::own<wasm::Module> module_;
+  wasm::own<wasm::Instance> instance_;
+  wasm::own<wasm::Memory> memory_;
+  wasm::own<wasm::Table> table_;
+
+  absl::flat_hash_map<std::string, wasm::own<wasm::Global>> host_globals_;
+  absl::flat_hash_map<std::string, FuncDataPtr> host_functions_;
+  absl::flat_hash_map<std::string, wasm::own<wasm::Func>> module_functions_;
+};
+
+// Helper functions.
+
+static std::string printValue(const wasm::Val& value) {
+  switch (value.kind()) {
+  case wasm::I32:
+    return std::to_string(value.get<uint32_t>());
+  case wasm::I64:
+    return std::to_string(value.get<uint64_t>());
+  case wasm::F32:
+    return std::to_string(value.get<float>());
+  case wasm::F64:
+    return std::to_string(value.get<double>());
+  default:
+    return "unknown";
+  }
+}
+
+static std::string printValues(const wasm::Val values[], size_t size) {
+  if (size == 0) {
+    return "";
+  }
+
+  std::string s;
+  for (size_t i = 0; i < size; i++) {
+    if (i) {
+      s.append(", ");
+    }
+    s.append(printValue(values[i]));
+  }
+  return s;
+}
+
+static const char* printValKind(wasm::ValKind kind) {
+  switch (kind) {
+  case wasm::I32:
+    return "i32";
+  case wasm::I64:
+    return "i64";
+  case wasm::F32:
+    return "f32";
+  case wasm::F64:
+    return "f64";
+  case wasm::ANYREF:
+    return "anyref";
+  case wasm::FUNCREF:
+    return "funcref";
+  default:
+    return "unknown";
+  }
+}
+
+static std::string printValTypes(const wasm::ownvec<wasm::ValType>& types) {
+  if (types.size() == 0) {
+    return "void";
+  }
+
+  std::string s;
```

Comment:
absl::StrJoin with Formatter should do this.

Reply:
It could, except that it doesn't know how to iterate over C-style arrays or `wasm::ownvec`, so I cannot use it without converting those to `std::array` or `std::vector`, which requires a copy, so I'd prefer not to do that.